R2API/ArtifactCodeAPI.cs

```diff
@@ -0,0 +1,192 @@
+using RoR2;
+using UnityEngine;
+using UnityEngine.Events;
+
+namespace R2API {
+
+    /// <summary>
+    /// API for adding custom artifact codes to the game.
+    /// </summary>
+    [R2APISubmodule]
+    public static class ArtifactCodeAPI {
+
+        private static readonly List<(ArtifactDef, Sha256HashAsset)> ArtifactsCodes = new List<(ArtifactDef, Sha256HashAsset)>();
+
+        /// <summary>
+        /// Return true if the submodule is loaded.
+        /// </summary>
+        public static bool Loaded {
+            get => _loaded;
+            internal set => _loaded = value;
+        }
+
+        private static bool _loaded;
+
+        #region Hooks
+        [R2APISubmoduleInit(Stage = InitStage.SetHooks)]
+        internal static void SetHooks() {
+            On.RoR2.PortalDialerController.Awake += AddCodes;
+            On.RoR2.PortalDialerController.PerformActionServer += PrintSha256HashCode;
+        }
+        /// <summary>
+        /// Prints the Artifact Code that the player inputs in the dialer. Useful for mod creators.
+        /// </summary>
+        /// <param name="orig"></param>
+        /// <param name="self"></param>
+        /// <param name="sequence"></param>
+        /// <returns></returns>
+        private static bool PrintSha256HashCode(On.RoR2.PortalDialerController.orig_PerformActionServer orig, PortalDialerController self, byte[] sequence) {
+            var result = self.GetResult(sequence);
+            R2API.Logger.LogInfo("Inputted Artifact Code:");
+            R2API.Logger.LogInfo("_00_07: " + result._00_07);
+            R2API.Logger.LogInfo("_08_15: " + result._08_15);
+            R2API.Logger.LogInfo("_16_23: " + result._16_23);
+            R2API.Logger.LogInfo("_24_31: " + result._24_31);
+            return orig(self, sequence);
+        }
+
+        /// <summary>
+        /// Adds custom ArtifactCodes to the portal dialer controller instance found in sky meadow.
+        /// </summary>
+        /// <param name="orig"></param>
+        /// <param name="self"></param>
+        private static void AddCodes(On.RoR2.PortalDialerController.orig_Awake orig, PortalDialerController self) {
+            foreach ((ArtifactDef, Sha256HashAsset) entry in ArtifactsCodes) {
+                var artifactDef = entry.Item1;
+                var artifactCode = entry.Item2;
+
+                if(!ArtifactCatalog.GetArtifactDef(artifactDef.artifactIndex)) {
+                    R2API.Logger.LogWarning($"ArtifactDef of name {artifactDef.cachedName} is not in the ArtifactCatalog. ignoring entry.");
+                    continue;
+                }
+
+                PortalDialerController.DialedAction dialedAction = new PortalDialerController.DialedAction();
+                dialedAction.hashAsset = artifactCode;
+                dialedAction.action = new UnityEvent();
+                dialedAction.action.AddListener(Wrapper);
+
+                void Wrapper() => self.OpenArtifactPortalServer(artifactDef);
```

Comment:
Can you seperate this from other lines of codes (at the end or the front) as it's kind of weird when it's behind the usage

Reply:
Might as well just use an anonymous delegate here

Reply:
I've made the requested change in the newest commit.

For future notice, the wrapper needs to be there due to the way how the UnityEvent class works in code.
Trying to simply do dialedAction.action.AddListener(self.OpenArtifactPortalServer(artifactDef) throws an error.